src/__spec__/interleave-generator.spec.ts

```diff
@@ -4,127 +4,83 @@ import * as iter from '../index'
 
-assert<
-  IterableIterator<string | number | Function>
->(
-  iter.interleaveGenerator(
-    function * (
-      canTakeAny: Function,
-      b1: iter.InterleaveBuffer<string>,
-      b2: iter.InterleaveBuffer<number>,
-      b3: iter.InterleaveBuffer<Function>
-    ) {
-      yield b1.take()
-      yield b2.take()
-      yield b3.take()
-    }
-  )(['foo'], [2], [(_: any) => _])
-)
+      for (const buffer of buffers) {
+        assert<string | undefined>(buffer.take())
+        assert<boolean>(buffer.canTake())
 
-assert<
-  IterableIterator<string | number | Function | {}>
->(
-  iter.interleaveGenerator(
-    function * (
-      canTakeAny: Function,
-      b1: iter.InterleaveBuffer<string>,
-      b2: iter.InterleaveBuffer<number>,
-      b3: iter.InterleaveBuffer<Function>,
-      b4: iter.InterleaveBuffer<{}>
-    ) {
-      yield b1.take()
-      yield b2.take()
-      yield b3.take()
-      yield b4.take()
+        if (buffer.canTake()) {
+          assert<iter.TakableInterleaveBuffer<string>>(buffer)
+          const res = buffer.take()
+          assert<string>(res)
+          yield res
+        }
+      }
     }
-  )(['foo'], [2], [(_: any) => _], [{}])
+  )(['foo'], ['bar'], ['baz'])
 )
 
 assert<
-  AsyncIterableIterator<string>
+  IterableIterator<string | number>
 >(
-  iter.asyncInterleaveGenerator(
-    async function * (
-      canTakeAny: Function,
-      buffer: iter.AsyncInterleaveBuffer<string>
+  iter.interleaveGenerator<string, number>(
+    function * gen (
+      canTakeAny,
+      b1,
+      b2
     ) {
-      yield await buffer.take()
-    }
-  )(['foo'])
-)
+      assert<iter.InterleaveCallback<[
+        iter.InterleaveBuffer<string>,
+        iter.InterleaveBuffer<number>
+      ]>>(gen)
+      assert<iter.InterleaveBuffer<string>>(b1)
+      assert<iter.InterleaveBuffer<number>>(b2)
 
-assert<
-  AsyncIterableIterator<string | number>
->(
-  iter.asyncInterleaveGenerator(
-    async function * (
-      canTakeAny: Function,
-      b1: iter.AsyncInterleaveBuffer<string>,
-      b2: iter.AsyncInterleaveBuffer<number>
-    ) {
-      yield await b1.take()
-      yield await b2.take()
-    }
-  )(['foo'], [2])
-)
+      if (b1.canTake()) {
+        assert<iter.TakableInterleaveBuffer<string>>(b1)
+        const res = b1.take()
+        assert<string>(res)
+        yield res
+      }
 
-assert<
-  AsyncIterableIterator<string | number | Function>
->(
-  iter.asyncInterleaveGenerator(
-    async function * (
-      canTakeAny: Function,
-      b1: iter.AsyncInterleaveBuffer<string>,
-      b2: iter.AsyncInterleaveBuffer<number>,
-      b3: iter.AsyncInterleaveBuffer<Function>
-    ) {
-      yield await b1.take()
-      yield await b2.take()
-      yield await b3.take()
+      if (b2.canTake()) {
+        assert<iter.TakableInterleaveBuffer<number>>(b2)
+        const res = b2.take()
+        assert<number>(res)
+        yield res
+      }
     }
-  )(['foo'], [2], [(_: any) => _])
+  )(['foo'], [2])
 )
-
-assert<
```

Comment:
Why would we not include any type test? We should test as much of type as we know works, even if some more complex type functionality does not yet.

Reply:
This is just glorified version of `assert<string, number>`, I don't see this as necessary.

Reply:
Huh? It's tests the async defs.

Reply:
> It tests the async defs.

Ah, then as I already said in the original comment, I have taken the async tests out until I touch it again.